File: sdk/go/sdk/buildkite/sdk.go
package buildkite

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/itchyny/json2yaml"
)

func NewPipeline() *Pipeline {
	return &Pipeline{}
}

type Pipeline struct {
	Agents map[string]any  `json:"agents,omitempty"`
	Env    map[string]any  `json:"env,omitempty"`
	Notify *PipelineNotify `json:"notify,omitempty"`
	Steps  []*PipelineStep `json:"steps"`
}

type pipelineStep interface {
	ToPipelineStep() *PipelineStep
}

func (p *Pipeline) AddStep(step pipelineStep) {
	p.Steps = append(p.Steps, step.ToPipelineStep())
}

func (p *Pipeline) AddAgent(key string, value any) {
	if p.Agents == nil {
		p.Agents = make(map[string]any)
	}

	p.Agents[key] = value
}

func (p *Pipeline) AddEnvironmentVariable(key string, value any) {
	if p.Env == nil {
		p.Env = make(map[string]any)
	}

	p.Env[key] = value
}

func (p *Pipeline) AddNotify(notify *PipelineNotify) {
	p.Notify = notify
}

func (p *Pipeline) ToJSON() (string, error) {
	data, err := json.MarshalIndent(p, "", "    ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func (p *Pipeline) ToYAML() (string, error) {
	data, err := p.ToJSON()
	if err != nil {
		return "", err
	}

	var output strings.Builder
	input := strings.NewReader(data)
	if err := json2yaml.Convert(&output, input); err != nil {
		return "", fmt.Errorf("converting JSON to YAML: %v", err)
	}

	return output.String(), nil
}

File: sdk/typescript/src/sdk.ts
import * as yaml from "yaml";
import * as schema from './types/schema'
export { EnvironmentVariable } from "./environment";

export class Pipeline {
    public agents: schema.AgentsObject = {};
    public env: schema.Env = {};
    public notify: schema.BuildNotify = [];
    public steps: schema.PipelineSteps = [];
    public secrets: schema.Secrets = [];

    /**
     * Set the pipeline
     * @param pipeline
     * @returns
     */
    setPipeline(pipeline: schema.BuildkitePipeline) {
        if (pipeline.agents) {
            this.agents = pipeline.agents
        }

        if (pipeline.env) {
            this.env = pipeline.env
        }

        if (pipeline.notify) {
            this.notify = pipeline.notify
        }

        if (pipeline.steps) {
            this.steps = pipeline.steps
        }
    }

    /**
     * Set the secrets for the pipeline
     * @param secrets
     * @returns
     */
    setSecrets(secrets: schema.Secrets) {
        this.secrets = secrets
    }

    /**
     * Add an agent to target by tag
     * @param tagName
     * @param tagValue
     * @returns
     */
    addAgent(tagName: string, tagValue: string) {
        this.agents[tagName] = tagValue;
    }

    /**
     * Add an environment variable
     * @param key
     * @param value
     */
    addEnvironmentVariable(key: string, value: any) {
        this.env[key] = value;
    }

    /**
     * Add an notification
     * @param notify
     */
    addNotify(notify: schema.BuildNotify[0]) {
        this.notify.push(notify);
    }

    /**
     * Add a step to the pipeline.
     * @param step
     * @returns
     */
    addStep(step: schema.PipelineSteps[0]) {
        this.steps.push(step);
        return this;
    }

    private build(): schema.BuildkitePipeline {
        const pipeline: schema.BuildkitePipeline = {};

        if (Object.keys(this.secrets).length > 0) {
            pipeline.secrets = this.secrets;
        }

        if (Object.keys(this.agents).length > 0) {
            pipeline.agents = this.agents;
        }

        if (Object.keys(this.env).length > 0) {
            pipeline.env = this.env;
        }

        if (Object.keys(this.notify).length > 0) {
            pipeline.notify = this.notify;
        }

        if (Object.keys(this.steps).length > 0) {
            pipeline.steps = this.steps;
        }

        return pipeline;
    }

    toJSON() {
        return JSON.stringify(this.build(), null, 4);
    }

    toYAML() {
        return yaml.stringify(this.build());
    }
}
